File: extract_DA.m
% extract_DA extracts duration (D) and cumulative amplitude (A) pairs from a
% 1D time series by segmenting based on sign changes in the derivative.

function [duration_sequence, amp_sequence,segments_idx] = extract_DA(time_series, lambda)

% INPUT:
%   time_series - 1D array representing the time series data.
%   lambda      - Scaling factor for thresholding.
%
% OUTPUT:
%   duration_sequence   - Array containing the duration (number of points) of each segment.
%   amp_sequence     - Array containing the cumulative amplitude (sum of values) of each segment.
if nargin<2
    lambda=0.001;
end
    % Step 1: Compute the raw derivative (no smoothing)
    derivative = diff(time_series);

    % Step 2: Compute an adaptive threshold using robust quantiles
    % This threshold is based on the difference between the 75th and 25th percentiles of diff(x)
    threshold = lambda * iqr(derivative);
    
    % Step 3: Segment the signal based on sign changes or low derivative values.
    % No minimum segment length is enforced.
    duration_sequence = [];
    amp_sequence = [];
    segments_idx=[];
    start_idx = 1;
    n = length(time_series);
    
    for i = 2:length(derivative)
        if sign(derivative(i)) ~= sign(derivative(i -1)) && (abs(derivative(i) - derivative(i-1))>threshold)
            % End the current segment
            duration = i - start_idx;
            amplitude = time_series(i) - time_series(start_idx);
            duration_sequence = [duration_sequence, duration];
            amp_sequence = [amp_sequence, amplitude];
            segments_idx = [segments_idx, i ];
            start_idx = i;  % Start a new segment
        end
    end


 
end

File: extract_DA.py
import numpy as np

def extract_DA(time_series, lambda_=0.001):
    """
    Extract duration (D) and cumulative amplitude (A) pairs from a 1D time series
    by segmenting based on sign changes in the raw derivative, with an adaptive
    threshold based on IQR(derivative).

    Parameters
    ----------
    time_series : array_like, shape (N,)
        1D time series.
    lambda_ : float, default=0.001
        Scaling factor for thresholding (threshold = lambda_ * IQR(diff(x))).

    Returns
    -------
    duration_sequence : np.ndarray, shape (M,)
        Duration (number of points) of each segment.
    amp_sequence : np.ndarray, shape (M,)
        Segment amplitude computed as x[i] - x[start_idx] at each cutpoint i.
    segments_idx : np.ndarray, shape (M,)
        Segment end indices i (1-based in MATLAB; here returned 0-based indices).
    """
    x = np.asarray(time_series, dtype=float).ravel()
    if x.size < 2:
        return np.array([], dtype=int), np.array([], dtype=float), np.array([], dtype=int)

    # Step 1: raw derivative (MATLAB diff)
    derivative = np.diff(x)

    # Step 2: adaptive threshold = lambda * IQR(derivative)
    q75, q25 = np.percentile(derivative, [75, 25])
    threshold = float(lambda_ * (q75 - q25))

    # Step 3: segment based on sign changes and jump size condition
    duration_list = []
    amp_list = []
    idx_list = []

    start_idx = 0  # Python 0-based; MATLAB start_idx=1
    # MATLAB: for i = 2:length(derivative)
    for i in range(1, derivative.size):
        if (np.sign(derivative[i]) != np.sign(derivative[i - 1])
            and (abs(derivative[i] - derivative[i - 1]) > threshold)):

            # MATLAB duration = i - start_idx; with MATLAB indices
            # Here, "i" is derivative index; corresponding signal index is i (0-based)
            duration = i - start_idx
            amplitude = x[i] - x[start_idx]

            duration_list.append(duration)
            amp_list.append(amplitude)
            idx_list.append(i)  # signal index where segment ends (0-based)

            start_idx = i

    return (np.asarray(duration_list, dtype=int),
            np.asarray(amp_list, dtype=float),
            np.asarray(idx_list, dtype=int))
